### ISSUE TITLE
Lasers Enigma compatibility

### ISSUE THREAD

**Issue by n0ct** [DRE2N/DungeonsXL#705]
Hello, I'm the developer of [lasers-enigma.eu](http://lasers-enigma.eu).

**Description**

A user reported to us that each of our plugin's components breaks when the instance of your plugin starts.

I don't really know what he means yet but my first guess is that you destroy every entity within the dungeon instance when it starts. Is that it?

**Comment by n0ct** [comment]
Probably ExistsAujourd’hui à 21:50
Howdy, I use a plugin that attempts to instance dungeons in Minecraft (dxl dungeons), and I've noticed all placed lasers etc. break when the instance starts. Would this be something to report to you guys, or to DXL?
 
bZxAujourd’hui à 22:12
@Probably Exists I'm unsure. what happens exactly ? Does the lasers and other components disappears but are lasers still visible ?
Does the barrier block (invisible solid block) placed on each component's location disappears too ?
Because I think (first guess) that DXL probably destroys every entities when the instance starts.
 
Probably ExistsAujourd’hui à 22:15
All of them are still placed, but they become non functioning
 
bZxAujourd’hui à 22:19
so they are visible but no lasers comes out of those ? Or does lasers still get out but you can't interact with them ?
 
Probably ExistsAujourd’hui à 22:20
Visible but no lasers/mirrors
 
bZxAujourd’hui à 22:20
hmm very strange. I have no idea of what DXL does. I will ask them directly.

**Comment by n0ct** [comment]
To be clear, a component (for examplea laser emitters) is composed of:
- A barrier block (the Minecraft invisible solid undestroyable block)
- One of more entities (invisible armor stand with visible custom heads)
- They can spawn particles that represent the light flow.

Knowing that how your plugin could interfere with mine?

**Comment by Sataniel98** [comment]
DungeonsXL doesn't kill entities, but it stores worlds separately from the world container and instantiates them when they are needed. That means, a dungeon world named "World" is saved as "World" in DXL's plugin folder, but can exist multiple times as a game instance under the name "DXL_Game_1", "DXL_Game_2" etc. if it is played parallelly by different players, or one time as "DXL_Edit_*" when it is edited. If your plugin saves data per world, then it will most likely save it e.g. for "DXL_Edit_X" where the laser is created by the builder and won't recognize the structures loaded to DXL_Game_Y.

tl;dr; This isn't your fault at all, DungeonsXL just isn't made to use non-Vanilla features out of the box.

**Comment by n0ct** [comment]
Ok thank you for your answer. I understand.

So in order to fix this issue, we have different solutions:
- Fix it in your plugin's code (I have an API you can use).
- Fix it in my plugin's code (Do you have an API I can use easily ? Events for instances creation/deletion/edition ? Will the compatibility be broken when you update your plugin ?).
- Do a little of both.

I think we should do a little of both in order to keep things separated properly:

- On my side:
 I add new methods to my API within my `Areas` class:
```
public void copyWorld(World originWorld, World destinationWorld, boolean clearAndReplace, boolean copyStatsAndPlayersData) throws IllegalStateException;
public void mergeStatsAndPlayersData(World originWorld, World destinationWorld) throws IllegalStateException;
public void clearWorld(World world) throws IllegalStateException;
```

- On your side, **whenever I will have published the new version containing those new methods** you will have to do the following:

1. Add my plugin within your pom: [you can find the required informations here](https://repository.lasers-enigma.eu/artifactory/webapp/#/artifacts/browse/tree/General/libs-release-local/eu/lasersenigma/LasersEnigma/6.6.0/LasersEnigma-6.6.0.jar)

2. Add it as an optional dependency within your plugin.yml

3. Add a small code within your plugin initialization that checks if my plugin is also loaded on the server your plugin is running on and store that information within a boolean accessible from other classes.

4. If that boolean value is true, call the methods I will have created (listed above) whenever you need to create instances (for edition or instance creation) or whenever you need to delete those instances.
I will be able to provide you detailed information about that if the methods prototypes are not already speaking for themselves.

Would you agree with this solution?

**Comment by Sataniel98** [comment]
Sounds reasonable. However I'm not sure about how to use the mergeStatsAndPlayersData method. Looks like you save stats per world, but I can't really use that since there's no world object to merge it with when the dungeon world unloads. I could just use the main world or the world where the player comes from but that would mess up the data of that world. An alternative could be to save data for other keys that aren't necessarily worlds sothat I can make it save the data under the name of the dungeon instead of the world, or not to save stats from dungeons at all.

**Comment by n0ct** [comment]
In reality, it is not entirely world by world. There is indeed a system of control points that saves data world by world but the majority of the data is saved puzzle area by puzzle area. Each puzzle area is itself specific to a world. The other data are linked to this area.

But you have highlighted a problem that will be complex to solve: Once data has been copied from one world to another, there is no longer any way to link the original data and the new data. Merging the data, therefore, becomes impossible. That means saving the progression of a player within an instance will require to code something to get around this problem.

So I have a question that will condition everything else:

Is it possible, with your plugin, to resume a dungeon at the same point where we stopped or is it that the instances are always fully reset (no backup of the progress of a player, location, inventory, ...)?

**Comment by n0ct** [comment]
I understand better what you meant by

> An alternative could be to save data for other keys that aren't necessarily worlds sothat I can make it save the data under the name of the dungeon instead of the world, or not to save stats from dungeons at all.

So the methods will be:
```java
public void copyWorld(String originWorldName, World destinationWorldName, boolean clearAndReplace, boolean copyStatsAndPlayersData) throws IllegalStateException;
public void mergeStatsAndPlayersData(String originWorldName, String destinationWorldName) throws IllegalStateException;
public void clearWorld(String worldName) throws IllegalStateException;
```
And those methods will be able to work on unloaded / unexisting worlds.

So you will have to do the following:

- When you create an edition session, before loading the world:
```java
copyWorld("DXL_DUNGEON_1", "DXL_EDIT_1", true, true); //the edit session can clear stats so it needs to get the previous stats
```
- When you end an edition session:
```java
copyWorld("DXL_EDIT_1", "DXL_DUNGEON_1", true, true); //the edit session can clear stats so it needs to erase the previous stats
```

- When you create an instance, before loading the world:
```java
copyWorld("DXL_DUNGEON_1", "DXL_GAME_1", true, true);
```

- When you end an instance:
```java
mergeStatsAndPlayersData("DXL_GAME_1", "DXL_DUNGEON_1");
clearWorld("DXL_GAME_1");
```
- When a dungeon is deleted:
```java
clearWorld("DXL_DUNGEON_1");
```



Some questions:
- Can you have multiple edit session of the same dungeon at the same time? I suppose you can't but it's just to be sure.
- Can you still play a dungeon while an edit session is ongoing? If so that's not a problem but that just means the player's stats and progress will not be saved in this case.
- The question I already asked in my previous comment: Is it possible, with your plugin, to resume a dungeon at the same point where we stopped or is it that the instances are always fully reset (no backup of the progress of a player, location, inventory, ...)?

**Comment by Sataniel98** [comment]
* Nope, 1 edit / dungeon map.
* Yes.
* Dungeon progress isn't persistent. It is possible that a player leaves the server and continues the game after whatever cooldown has been set up for him to be kicked, but the progress isn't saved to disc. Backups of instances aren't made.

**Comment by n0ct** [comment]
Ok so I will do what's required on my side and let you know when it's ready.

**Comment by Sataniel98** [comment]
I'll close this for now. If you're still interested in realizing the integration, please let me know.

**Comment by n0ct** [comment]
Hello @Sataniel98 ,
I'm back on my plugin development and I implemented the required methods (not exactly as we planed earlier but, hopefully, enough to make it work).

Could you please do the following modifications:

* Add the lasers-enigma maven repository:
```xml
        <repository>
            <id>lasers-enigma</id>
            <url>https://repository.lasers-enigma.eu/artifactory/r/</url>
        </repository>
```
* Add the dependency for the latest snapshot:
```xml
        <dependency>
            <groupId>eu.lasersenigma</groupId>
            <artifactId>LasersEnigma</artifactId>
            <version>6.9.0-20201030.182456-2</version>
            <scope>compile</scope>
        </dependency>
```
[see here for more informations](https://repository.lasers-enigma.eu/artifactory/webapp/#/artifacts/browse/tree/General/r/eu/lasersenigma/LasersEnigma/6.9.0-SNAPSHOT/LasersEnigma-6.9.0-20201030.182456-2.jar)

* Include my plugin "LasersEnigma" in your plugin.yml softdepend:
```yml
softdepend: [LasersEnigma]
```

* For each next instructions, please check if my plugin is available on the server first:
```java
if (getServer().getPluginManager().isPluginEnabled("LasersEnigma")) {
   //...
}
```

* On INSTANCE / EDIT session creation. **This must be done BEFORE the world is loaded**:
```java
Areas.getInstance().copyWorldData("MyDungeonName", "MyInstanceOrEditSessionWorldName", "true");
```

* On EDIT session save.
```java
Areas.getInstance().copyWorldData("MyEditSessionWorldName", "MyDungeonName", "true");
```

* On INSTANCE / EDIT session deletion. **This must be done AFTER the world is unloaded**:
```java
Areas.getInstance().removeWorldData("MyInstanceWorldName");
```

Do not hesitate to contact me if you encounter any problems.

Thank you per advance 😃

**Comment by n0ct** [comment]
I saw that you had developed an API and events that I could have listened to. But I'm not sure that your events are called before the world loads / after the world unloads ... ?

**Comment by Sataniel98** [comment]
If you'd like to use our API, you're very welcome to do so of course.

Before the world loads: [ResourceWorldInstantiateEvent](https://erethon.de/javadocs/dungeonsxl/de/erethon/dungeonsxl/api/event/world/ResourceWorldInstantiateEvent.html)

When the edit world is saved: [EditWorldSaveEvent](https://erethon.de/javadocs/dungeonsxl/de/erethon/dungeonsxl/api/event/world/EditWorldSaveEvent.html)

I've also added an event that fires after the instance is unloaded: [InstanceWorldPostUnloadEvent](https://erethon.de/javadocs/dungeonsxl/de/erethon/dungeonsxl/api/event/world/InstanceWorldPostUnloadEvent.html) Make sure to [update](https://erethon.de/repo/de/erethon/dungeonsxl/dungeonsxl-api/) the dependency when you use this.

As a rule of thumb, you can tell whether an event fires for or after the action from if it implements Cancellable or not. If an event is cancellable, it must fire beforehand.

**Comment by n0ct** [comment]
## Checksums (not an issue)
It seems no checksum is available for the API within your repository:
I had to define your repository in my pom using:

```xml
        <repository>
            <id>dre-repo</id>
            <url>http://erethon.de/repo</url>
            <releases>
                <enabled>true</enabled>
                <updatePolicy>always</updatePolicy>
                <checksumPolicy>warn</checksumPolicy>
            </releases>
        </repository>
```
If you don't plan to improve your repository by adding checksums, maybe you should plan to add this "release" node within your API documentation.

## Where are worlds and dungeons names ?

I'm not sure to understand where to find the names of the bukkit worlds / dungeons within each of your events:

Could you please check the following code in order for me to be sure I didn't do mistakes ?
I would like to be sure that the names I use within my methods calls are always the name of the dungeon or the name of the bukkit world corresponding to this game instance / edit session.

#### ResourceWorldInstantiateDungeonsXLListener.java
```java
    /**
     * executed on a DungeonsXL ResourceWorldInstantiateEvent
     * On INSTANCE / EDIT session creation. This is called BEFORE the world is loaded.
     *
     * @param e a ResourceWorldInstantiateEvent
     */
    public void onResourceWorldInstantiateEvent(ResourceWorldInstantiateEvent e) {
        Main.getInstance().getBetterLogger().log(Level.FINEST, "onResourceWorldInstantiateEvent");
        Areas.getInstance().copyWorldData(e.getResource().getName(), e.getResource().getEditWorld().getName(), true);
    }
```
#### EditWorldSaveDungeonsXLListener.java
```java
    /**
     * executed on a DungeonsXL EditWorldSaveEvent
     * On EDIT session save.
     *
     * @param e a EditWorldSaveEvent
     */
    public void onEditWorldSaveEvent(EditWorldSaveEvent e) {
        Main.getInstance().getBetterLogger().log(Level.FINEST, "onEditWorldSaveEvent");
        Areas.getInstance().copyWorldData(e.getInstance().getWorld().getName(), e.getEditWorld().getResource().getName(), true);
    }
```
#### InstanceWorldPostUnloadEventDungeonsXLListener.java
```java
    /**
     * executed on a DungeonsXL InstanceWorldPostUnloadEvent
     * On INSTANCE / EDIT session deletion. This is called AFTER the world is unloaded.
     *
     * @param e a InstanceWorldPostUnloadEvent
     */
    public void onInstanceWorldPostUnloadEvent(InstanceWorldPostUnloadEvent e) {
        Main.getInstance().getBetterLogger().log(Level.FINEST, "onInstanceWorldPostUnloadEvent");
        Areas.getInstance().removeWorldData(e.getInstanceWorldName());
    }
```

[The complete branch can be seen here](https://gitlab.com/lasersenigma/lasersenigma/-/compare/master...features%2FDungeonsXL-compatibility).

Thank you.

**Comment by Sataniel98** [comment]
InstanceWorld#getName() returns the name of the resource world. If the world is called DungeonMap and as an instance DXL_Edit_1 and DXL_Game_2, what this does is the following:

1st copies world data from DungeonMap to DungeonMap (seems I'll have to look into this to make a possibility to get the new world's name, since it's not loaded at that point)

2nd copies from DXL_Edit_1 to DungeonMap (though e.getEditWorld().getName() would have been fine)

3rd removes world data from DXL_Edit_1 or DXL_Game_2

As for the checksums, I don't think I need them because I don't usually have snapshots uploaded there. This time is an exception because no stable release of the API module exists.

**Comment by n0ct** [comment]
Ok. Thank you. Indeed I would need you to look into a way to give me the new world's name before it's loaded.

**Comment by Sataniel98** [comment]
I've added ResourceWorldInstantiateEvent#getInstanceWorldName() https://erethon.de/javadocs/dungeonsxl/de/erethon/dungeonsxl/api/event/world/ResourceWorldInstantiateEvent.html#getInstanceWorldName--

Hope that helps.

**Comment by n0ct** [comment]
Thank you, it will indeed help. I will finish the code accordingly.

Have a good evening.

**Comment by Sataniel98** [comment]
Glad to hear, please let me know if something doesn't work. Have a nice evening, too :)

**Comment by n0ct** [comment]
I found a strange behavior using dungeonsxl-dist-0.18-SNAPSHOT-1063.jar:

My test case is as follows:
* /dxl edit plop
* Create a lasers puzzle
* /dxl save
* Check that the created lasers puzzle still exist
* /dxl leave
* /dxl play plop
* Check that the created lasers puzzle still exist

I found issues:
* /dxl edit plop => ResourceWorldInstantiateEvent => Not called.
* /dxl save => EditWorldSaveEvent => Called correctly
  BUT then it reloads a new edit session with a different world name and does not call ResourceWorldInstantiateEvent.
* /dxl leave => onInstanceWorldPostUnloadEvent => Called correctly.
* /dxl play plop => onResourceWorldInstantiateEvent => Not called.

**Comment by Sataniel98** [comment]
Alright, I'll have to test that.

**Comment by n0ct** [comment]
If you need help at some point don't hesitate to ask here or on discord.

**Comment by Sataniel98** [comment]
Thanks, I'll get back to you when I've worked on it. It might take a bit since the semester has started again.

**Comment by n0ct** [comment]
Sorry to ask again but some users are waiting for this feature. Could you tell me when you think you will be able to check / do this fix ?

**Comment by Sataniel98** [comment]
I can't give an ETA unfortunately.

**Comment by n0ct** [comment]
Ok so I will tell my users the same ^^

**Comment by Sataniel98** [comment]
Should work now.

**Comment by n0ct** [comment]
According to one of my users, it seems there is still a problem. I did not took the time to investigate yet.

**Comment by Sataniel98** [comment]
Please let me know if you find something.

**Comment by n0ct** [comment]
I found the first issue (there may be other issues after this one):

```[22:33:34] [Server thread/INFO]: bZx_ issued server command: /dxl save
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] onEditWorldSaveEvent
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] copyWorldData(originWorldName: DXL_Edit_0, destinationWorldName: test, clearAndReplace: true)
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] Database.clearWorldData(worldName: test)
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing playersworldcheckpoint data for this world.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing checkpoints from checkpoints index (playerCheckpoint).
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing playerStats.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing playerKeys.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing scheduled actions.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing linked components.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing other components.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing victory areas.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Removing areas.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] Database.copyWorldData(originWorldName: DXL_Edit_0, destinationWorldName: test)
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Copying areas.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Copying victory areas.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Copying unlinked components.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Copying linked components.
[22:33:34] [Server thread/INFO]: [LasersEnigma] [FINEST] |- Copying scheduled actions.
[22:33:35] [Server thread/WARN]: Entity EntityPlayer['bZx_'/135, l='ServerLevel[world]', x=109.14, y=21.00, z=-335.65](bZx_ at 109.13614420775649,21.0,-335.64695204279207) left loaded chunk area
[22:33:35] [Server thread/WARN]: [Multiverse-Core] Hmm Multiverse does not know about this world but it's loaded in memory.
[22:33:35] [Server thread/WARN]: [Multiverse-Core] To let Multiverse know about it, use:
[22:33:35] [Server thread/WARN]: [Multiverse-Core] /mv import DXL_Edit_0 NORMAL
[22:33:35] [Server thread/INFO]: [LasersEnigma] [FINEST] onWorldLoadEvent
[22:33:35] [Server thread/INFO]: [LasersEnigma] [FINER] Areas.onWorldUnload(DXL_Edit_0)
[22:33:35] [Server thread/INFO]: ThreadedAnvilChunkStorage (DXL_Edit_0): All chunks are saved
[22:33:39] [Server thread/INFO]: -------- World Settings For [DXL_Edit_1] --------
[22:33:39] [Server thread/INFO]: Entity Tracking Range: Pl 48 / An 48 / Mo 48 / Mi 32 / Other 64
[22:33:39] [Server thread/INFO]: View Distance: 10
[22:33:39] [Server thread/INFO]: Item Merge Radius: 2.5
[22:33:39] [Server thread/INFO]: Item Despawn Rate: 6000
[22:33:39] [Server thread/INFO]: Allow Zombie Pigmen to spawn from portal blocks: true
[22:33:39] [Server thread/INFO]: Zombie Aggressive Towards Villager: true
[22:33:39] [Server thread/INFO]: Arrow Despawn Rate: 1200 Trident Respawn Rate:1200
[22:33:39] [Server thread/INFO]: Nerfing mobs spawned from spawners: false
[22:33:39] [Server thread/INFO]: Mob Spawn Range: 6
[22:33:39] [Server thread/INFO]: Custom Map Seeds:  Village: 10387312 Desert: 14357617 Igloo: 14357618 Jungle: 14357619 Swamp: 14357620 Monument: 10387313 Ocean: 14357621 Shipwreck: 165745295 End City: 10387313 Slime: 987234911 Bastion: 30084232 Fortress: 30084232 Mansion: 10387319 Fossil: 14357921 Portal: 34222645
[22:33:39] [Server thread/INFO]: Cactus Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Cane Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Melon Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Mushroom Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Pumpkin Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Sapling Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Beetroot Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Carrot Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Potato Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Wheat Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: NetherWart Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Vine Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Cocoa Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Bamboo Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: SweetBerry Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Kelp Growth Modifier: 100%
[22:33:39] [Server thread/INFO]: Entity Activation Range: An 32 / Mo 32 / Ra 48 / Mi 16 / Tiv true
[22:33:39] [Server thread/INFO]: Hopper Transfer: 8 Hopper Check: 1 Hopper Amount: 1
[22:33:39] [Server thread/INFO]: Experience Merge Radius: 3.0
[22:33:39] [Server thread/INFO]: Max TNT Explosions: 100
[22:33:39] [Server thread/INFO]: Tile Max Tick Time: 50ms Entity max Tick Time: 50ms
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] onWorldLoadEvent
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINER] Areas.onWorldLoad(DXL_Edit_1)
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] PlayerInventorySaveManager.addInventoryFromDatabase
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] PlayerInventorySaveManager.addInventoryFromDatabase
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] PlayerInventorySaveManager.addInventoryFromDatabase
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] PlayerInventorySaveManager.addInventoryFromDatabase
[22:33:39] [Server thread/INFO]: [LasersEnigma] [FINEST] PlayerInventorySaveManager.addInventoryFromDatabase
[22:33:39] [Server thread/WARN]: Entity EntityPlayer['bZx_'/135, l='ServerLevel[DXL_Edit_1]', x=-68.93, y=4.00, z=141.71](bZx_ at -68.9291998726164,4.0,141.71348466732258) left loaded chunk area
```

The error can be very easily reproduced using the following test case:
- /dxl create test
- Create a lasers area and just put a laser sender inside it
- /dxl save

So the issue can be seen within the log when I write /dxl save (dungeon name: test, edit session: DXL_Edit_0):

- EditWorldSaveEvent is correctly called:
My plugin's data related to the world "test" are cleared (it's useless in my test case. but this can be useful if there is previous data for this dungeon).
My plugin's data are copied from world "DXL_Edit_0" to the world "test".

That part FINE.

- The player is teleported out of the world DXL_Edit_0

- After a short duration your plugin start a new edit session named "DXL_Edit_1".

- **The event "ResourceWorldInstantiateEvent" was NOT called (so my plugin's data were not copied from "test" to "DXL_EDIT_1").**

- The player is teleported in the world DXL_Edit_1

- The lasers we just created do not work anymore.

**Comment by n0ct** [comment]
Could you please try it by yourself next time. This will save us from having to wait for each other. You will be able to see directly if your fix is working.
The latest version of my plugin is available here: [6.11.1](http://lasers-enigma.eu/en/the-plugin/download/)

**Comment by n0ct** [comment]
within my plugin config file you could select the log level: FINEST
This could help you see when your events are called.
Btw if you need any help or explanation I'm available.

**Comment by n0ct** [comment]
I come back to you hoping for some news. Sorry to ask again but some users asked me what's going on again. They would really be happy to see this compatibility work. Don't hesitate to tell me if I can help.

**Comment by Sataniel98** [comment]
I was sure I had replied about a week ago, perhaps I forgot to click the comment button... Sorry for that! My debug plugin says the event fires as it should upon /dxl save since the fix. Perhaps you weren't using the latest build?

https://erethon.de/repo/de/erethon/dungeonsxl/dungeonsxl-dist/0.18/dungeonsxl-dist-0.18-SNAPSHOT-1070.jar

**Comment by n0ct** [comment]
I tested it. Indeed I didn't have the latest dev build. It works.
My users will be very happy to see that compatibility finally working ! Thank you very much.

Have a good evening.
